### DISCUSSION TITLE
How to build static binaries under linux, not shared?

### DISCUSSION THREAD

**Discussion by KarelWintersky** [SoftEtherVPN/SoftEtherVPN#2057] [General]
### Are you using SoftEther VPN 5.x?

- [X] Yes, I'm using SoftEther VPN 5.x, not 4.x.

### Component

VPN Server, VPN Bridge, VPN Client, VPN Tools

### Operating system & version

Linux

### Architecture or Hardware model

amd64

### Steps to reproduce

How to build static binaries under linux, not shared?

### ✔️ Expected Behavior

static binaries

### ❌ Actual Behavior

shared binaries

**Comment by siddharth-narayan** [comment]
It took a while, but I was able to build statically by following these steps:

Insert this code into ```src/CMakeLists.txt``` somewhere where it will affect your build (I chose in the UNIX if block):
```
set(CMAKE_FIND_LIBRARY_SUFFIXES .a)
set(BUILD_SHARED_LIBS OFF)
set(CMAKE_EXE_LINKER_FLAGS "-static")
```

This will essentially tell CMake to build everything statically, and only look for static libraries.

In ```src/Cedar/CMakeLists.txt:12``` > ```add_library(cedar SHARED ${SOURCES_CEDAR} ${SOURCES_CEDAR_CPP} ${HEADERS_CEDAR})```
Change "SHARED" to "STATIC"

Do the same for ```src/Mayaqua/CMakeLists.txt:7```

Make sure, of course to have all the dependencies available as static libraries. You might have some trouble with statically linking libc. I personally got it to work with musl.

Good Luck!

**Comment by siddharth-narayan** [comment]
I forgot to mention that you also have to switch the order in which curses and readline are linked for the cedar target (in cedar's CMakeLists.txt), otherwise you will get undefined reference errors.

**Comment by chipitsine** [comment]
I thought that "cmake -DBUILD_SHARED_LIBS=OFF ...." should help, but I didn't try yet. also, there might be caveats with libs as mentioned in the previous message

if there's high demand, we can add "-DBuild_SHARED_LIBS=OFF" to CI matrix, not a big deal

(editing CMakeLists.txt is weird approach, I think overriding from command line is more straightforward)

**Reply by siddharth-narayan** [reply]
> if there's high demand, we can add "-DBuild_SHARED_LIBS=OFF" to CI matrix, not a big deal

> (editing CMakeLists.txt is weird approach, I think overriding from command line is more straightforward)

I don't think building statically is possible from commandline. As far as I know, setting BUILD_SHARED_LIBS to off doesn't make any difference because it's explicitly overriden here:
https://github.com/SoftEtherVPN/SoftEtherVPN/blob/5d1ce1a2cd3e3681ff60f52db42a1b98280e0162/src/Cedar/CMakeLists.txt#L12 and here:
https://github.com/SoftEtherVPN/SoftEtherVPN/blob/5d1ce1a2cd3e3681ff60f52db42a1b98280e0162/src/Mayaqua/CMakeLists.txt#L7 Also, because readline is specified after ncurses, this fails when static linking because in static form, readline depends on ncurses but the linker isn't able to resolve symbols when ncurses is specified first:
https://github.com/SoftEtherVPN/SoftEtherVPN/blob/5d1ce1a2cd3e3681ff60f52db42a1b98280e0162/src/Cedar/CMakeLists.txt#L68

**Comment by KarelWintersky** [comment]
I can't say that this task is needed by a wide range of users, but in my organization there was a need (or rather, extreme desirability) to build this package not in shared, but in static mode. I tried to understand how it could be done myself, but my knowledge of C leaves much to be desired.

Which prompted the question.

The thing is that the [softether download center](https://www.softether-download.com/en.aspx) offers to download archives with templates for compilation for a specific platform, with the help of which static binaries are built.

But it seems to me that it is better to build the current release from git than packages that are more than a year out of date.

Thank you.

**Comment by chipitsine** [comment]
@siddharth-narayan  you may be interested in 

```
option(OQS_PROVIDER_BUILD_STATIC "Build a static library instead of a shared library" OFF)
if(OQS_PROVIDER_BUILD_STATIC AND BUILD_SHARED_LIBS)
  message(FATAL_ERROR "`OQS_PROVIDER_BUILD_STATIC` is not compatible with `BUILD_SHARED_LIBS`.")
endif()
```

(found in oqs provider). my previous experiments with OpenSSL engines shown me that they do not like static linking.

as for the rest, I've changed (according to your findings), at least Cedar and Mayaqua are easily link statically.
I'm looking at the rest

**Comment by siddharth-narayan** [comment]
Sorry, I'm a little confused. I ***was*** able to build statically, including OpenSSL. It's just not possible through commandline, because of the reasons I mentioned. OQSProvider is always built statically, because it should be included even on Windows, where it's very difficult to link dynamically. Are you still trying to compile statically?

**Reply by chipitsine** [reply]
nevermind, it might be my previous experience.

**Comment by chipitsine** [comment]
@siddharth-narayan 

I tried the following patch

```diff
diff --git a/src/Cedar/CMakeLists.txt b/src/Cedar/CMakeLists.txt
index 7818d790..b525522b 100644
--- a/src/Cedar/CMakeLists.txt
+++ b/src/Cedar/CMakeLists.txt
@@ -9,7 +9,7 @@ endif()
 if(WIN32)
   add_library(cedar STATIC ${SOURCES_CEDAR} ${SOURCES_CEDAR_CPP} ${HEADERS_CEDAR})
 else()
-  add_library(cedar SHARED ${SOURCES_CEDAR} ${SOURCES_CEDAR_CPP} ${HEADERS_CEDAR})
+  add_library(cedar ${SOURCES_CEDAR} ${SOURCES_CEDAR_CPP} ${HEADERS_CEDAR})
 endif()
 
 set_target_properties(cedar
diff --git a/src/Mayaqua/CMakeLists.txt b/src/Mayaqua/CMakeLists.txt
index 51acec0b..d18f2e2a 100644
--- a/src/Mayaqua/CMakeLists.txt
+++ b/src/Mayaqua/CMakeLists.txt
@@ -4,7 +4,7 @@ file(GLOB HEADERS_MAYAQUA "*.h" "Crypto/*.h")
 if(WIN32)
   add_library(mayaqua STATIC ${SOURCES_MAYAQUA} ${HEADERS_MAYAQUA})
 else()
-  add_library(mayaqua SHARED ${SOURCES_MAYAQUA} ${HEADERS_MAYAQUA})
+  add_library(mayaqua ${SOURCES_MAYAQUA} ${HEADERS_MAYAQUA})
 endif()
 
 target_include_directories(mayaqua PUBLIC .)

```

that part works as expected, i.e. Cedar and Mayaqua are linked either statically (if `-DBUILD_SHARED_LIBS=OFF` is passed) or dynamically. there are may be issues with "make install", but I did not test yet.

as for the rest, I gave up on convincing "find_package" that it should pick static lib. not sure whether that's possible.

what do you think, shall we merge the patch above ?

**Comment by siddharth-narayan** [comment]
As I mentioned, there's also the problem of the readline library being "after" ncurses, because it means the linker fails to resolve symbols correctly (which is ridiculous). To get find_package to pick static libraries, you can do this: ```set(CMAKE_FIND_LIBRARY_SUFFIXES .a)```

I also included this (I found it on a stackoverflow question about converting dynamically linked to statically linked)
```set(CMAKE_EXE_LINKER_FLAGS "-static")```

So here's what my git diff looks like (very minor changes from yours):
```
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index c49a3c78..30df30a7 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -81,6 +81,9 @@ if(UNIX)
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fsigned-char")
   set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -g")
   set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2")
+  set(CMAKE_FIND_LIBRARY_SUFFIXES .a)
+  set(BUILD_SHARED_LIBS OFF)
+  set(CMAKE_EXE_LINKER_FLAGS "-static")
 
   add_definitions(-DUNIX -DOS_UNIX)
 
diff --git a/src/Cedar/CMakeLists.txt b/src/Cedar/CMakeLists.txt
index 7818d790..3fb80dec 100644
--- a/src/Cedar/CMakeLists.txt
+++ b/src/Cedar/CMakeLists.txt
@@ -9,7 +9,7 @@ endif()
 if(WIN32)
   add_library(cedar STATIC ${SOURCES_CEDAR} ${SOURCES_CEDAR_CPP} ${HEADERS_CEDAR})
 else()
-  add_library(cedar SHARED ${SOURCES_CEDAR} ${SOURCES_CEDAR_CPP} ${HEADERS_CEDAR})
+  add_library(cedar ${SOURCES_CEDAR} ${SOURCES_CEDAR_CPP} ${HEADERS_CEDAR})
 endif()
 
 set_target_properties(cedar
@@ -65,7 +65,7 @@ if(UNIX)
   find_package(Curses REQUIRED)
   find_library(LIB_READLINE readline)
 
-  target_link_libraries(cedar PRIVATE ${CURSES_LIBRARIES} ${LIB_READLINE})
+  target_link_libraries(cedar PRIVATE ${LIB_READLINE} ${CURSES_LIBRARIES})
 
   if(${CMAKE_SYSTEM_NAME} STREQUAL "Darwin")
     target_link_libraries(cedar PRIVATE mayaqua pcap)
diff --git a/src/Mayaqua/CMakeLists.txt b/src/Mayaqua/CMakeLists.txt
index 51acec0b..b0f559d8 100644
--- a/src/Mayaqua/CMakeLists.txt
+++ b/src/Mayaqua/CMakeLists.txt
@@ -4,7 +4,7 @@ file(GLOB HEADERS_MAYAQUA "*.h" "Crypto/*.h")
 if(WIN32)
   add_library(mayaqua STATIC ${SOURCES_MAYAQUA} ${HEADERS_MAYAQUA})
 else()
-  add_library(mayaqua SHARED ${SOURCES_MAYAQUA} ${HEADERS_MAYAQUA})
+add_library(mayaqua ${SOURCES_MAYAQUA} ${HEADERS_MAYAQUA})
 endif()
 
 target_include_directories(mayaqua PUBLIC .)
@@ -100,7 +100,7 @@ if(UNIX)
 
   find_library(LIB_M m)
   find_library(LIB_RT rt)
-
+  
   target_link_libraries(mayaqua
     PRIVATE
       Threads::Threads
```

**Reply by chipitsine** [reply]
interesting. `set(CMAKE_FIND_LIBRARY_SUFFIXES .a)` can help. I'll try tomorrow, thanks!
